gesturelibrary/src/double.c

```diff
@@ -41,45 +58,199 @@ int set_on_double_tap(void (*listener)(const double_tap_t*)) {
+                // printf("Printing tap data: x = %f\n", tap->x);
+                // printf("Printing distances: x_diff = %f, y_diff = %f, time_diff = %f\n", x_diff, y_diff, time_diff);
+
+                if (x_diff < DOUBLE_MAX_X && y_diff < DOUBLE_MAX_Y) {
+                    if (time_diff < DOUBLE_TIME_DIFF) {
+                        // found the corresponding d_tap
+                        double_tap_d[i].state = RECOGNIZER_STATE_COMPLETED;
+                        null_index            = -1;
+                        update                = 0;
+
+                        // update data fields
+                        double_tap_d[i].x = event->x;
+                        double_tap_d[i].y = event->y;
+                        double_tap_d[i].t = event->t;
+                    } else {
+                        double_tap_d[i].state = RECOGNIZER_STATE_FAILED;
+                    }
+
+                    if (on_double_tap) {  // listener
+                        on_double_tap(&double_tap_d[i]);
+                    }
+                } else {  // grouping method needs to change
+                    float group_time_diff = (tap->t - d_tap.t0) < 0 ? (tap->t - d_tap.t0) * -1 : (tap->t - d_tap.t0);
+
+                    if (group_time_diff <
+                        DOUBLE_GROUP_TIME_DIFF) {  // if not it, we need to check if this is part of the same group
+                        group_index = i;
+                        // group_number = d_tap.group; //set this one's group to this if group_index dne -1
+                    }
+                }
+
+                break;
+            case RECOGNIZER_STATE_FAILED:
+            case RECOGNIZER_STATE_NULL:
+                // printf("entered null with tap x = %f\n", tap->x);
+                // save the index (need to replace)
+                null_index = i;  // we want to prioritize null spots and failed spots over completed ones
+                break;
+            case RECOGNIZER_STATE_COMPLETED:
+                complete_index = i;
+            default:
+                break;
             }
         }
-        if (on_double_tap) {
-            on_double_tap(double_tap);
-        }
-        break;
-    case RECOGNIZER_STATE_FAILED:
-    case RECOGNIZER_STATE_NULL:
-    case RECOGNIZER_STATE_COMPLETED:
-        if (tap->state == RECOGNIZER_STATE_COMPLETED) {
-            double_tap->state = RECOGNIZER_STATE_IN_PROGRESS;
+
+        if (update) {
+            if (null_index != -1) {
+                // printf("entered null_index if with tap x = %f and index = %d\n", tap->x, null_index);
+                update_data0_fields(null_index, group_index, event);
+            } else if (complete_index != -1) {
+                update_data0_fields(complete_index, group_index, event);
+            }
         }
-        if (on_double_tap) {
-            on_double_tap(double_tap);
+
+        for (int j = 0; j < MAX_TOUCHES; j++) {
+            printf("Printing state: %d, x: %f, x0: %f, group: %d\n",
```

Comment:
Please remove or comment debugging statements.
Your file should not include `stdio.h` when you are done.